### PR TITLE
View / apply constraints when an interaction starts

### LINKED ISSUE #9396: Unexpected animation when starting to interact with the map
**Describe the bug**
This issue is similar to #9366. When starting to interact with a map that has a view whose zoom and center were set with `View#setZoom()` and `View#setCenter()` or `View#animate()` and are outside the restricted range, an unexpected zoom-out occurs.

**To Reproduce**
1. Go to https://codesandbox.io/s/3n2o1n5qm (uses `View#setZoom()` and `View#setCenter()` or https://codesandbox.io/s/o51l6l6qrz (uses `View#animate()`)
2. Pan the map
3. See the map zoom out

The minimal map configuration to reproduce this is:
```js
const map = new Map({
  layers: [new TileLayer({ source: new OSM() })],
  target: 'map'
});
const view = map.getView();
view.animate({
  center: [0, 0],
  zoom: 0
});
// or, instead of view.animate:
//view.setCenter([0, 0]),
//view.setZoom(0);
```

The same happens when the map is initially already configured with a view with `zoom` and `center`.

**Expected behavior**
Panning the map with the above setup does not lead to an unexpected zoom-out.

### PR DESCRIPTION
**Pull request by jahow**
Previously, an interaction could begin while target values (center/resolution) were out of the allowed range, causing a glitch where the view zoom/position would jump suddenly.

@ahocevar I've reused your test from #9399 & expanded on it.

Fixes #9396.

### CODE REVIEW

File: src/ol/control/Zoom.js [hunk 1]
```diff
@@ -114,6 +114,7 @@ class Zoom extends Control {
       // upon it
       return;
     }
+    view.resolveConstraints(0);
```
Comment (ahocevar) [inline src/ol/control/Zoom.js]: I'm wondering if a better place for this new line this would be ´ol/View#animate()`, before the last line.

File: src/ol/interaction/Interaction.js [hunk 2]
```diff
@@ -135,6 +135,7 @@ export function zoomByDelta(view, delta, opt_anchor, opt_duration) {
   const newZoom = view.getConstrainedZoom(currentZoom + delta);
   const newResolution = view.getResolutionForZoom(newZoom);
 
+  view.resolveConstraints(0);
```
Comment (ahocevar) [inline src/ol/interaction/Interaction.js]: Is this needed? Or will it still be needed if `ol/View#animate()` handles it (like I suggested in the previous comment)?

File: test/spec/ol/view.test.js [hunk 3]
```diff
@@ -1801,6 +1801,52 @@ describe('ol.View', function() {
...
+  beforeEach(function() {
+    map = new Map({
+      target: createMapDiv(512, 256)
+    });
+  });
+  afterEach(function() {
+    disposeMap(map);
+  });
+
+  it.only('works when initialized with #setCenter() and #setZoom()', function(done) {
```
Comment (ahocevar) [inline test/spec/ol/view.test.js]: The `.only` should be removed.

File: test/spec/ol/view.test.js [hunk 4]
```diff
@@ -1801,6 +1801,52 @@ describe('ol.View', function() {
...
+    view.setZoom(0);
+    view.beginInteraction();
+    view.endInteraction();
+    setTimeout(function() {
+      expect(callCount).to.be(1);
+      done();
+    }, 500);
+  });
+
+  it.only('works when initialized with #animate()', function(done) {
```
Comment (ahocevar) [inline test/spec/ol/view.test.js]: `.only` also should be removed here.

File: src/ol/control/Zoom.js (hunk 1 above)
Comment (jahow) [reply src/ol/control/Zoom.js]: I've thought about that as well but didn't manage to get it to work: it was either breaking the "smoothly return view in allowed extent/resolution" effect at the end of an interaction, or causing regressions in other places, depending on where I'd put it.

I agree this does not feel right. To have a cleaner system would mean some rework of the `animate` method IMO, which is a tricky task. The ideal solution would be to get rid of all these `applyTargetState_` and `resolveConstraints` calls and only translate target values into actual view parameters on the main render loop (ans as such apply constraints only once per frame).

File: test/spec/ol/view.test.js (hunk 3 above)
Comment (jahow) [reply test/spec/ol/view.test.js]: fixed

File: src/ol/interaction/Interaction.js (hunk 2 above)
Comment (jahow) [reply src/ol/interaction/Interaction.js]: See other comment

**Comment by jahow** [comment]
TL;DR: could not find a better way with the time & means available to me, sorry!

**Comment by ahocevar** [comment]
@jahow I finally wrapped my head around the new constraints system, and I think I came up with a good solution. Can you please let me know what you think about 070c1ec?

**Comment by jahow** [comment]
Works fine! well done
